src/community/gdal/gdal-wcs/src/main/java/org/geoserver/wcs/response/GdalCoverageResponseDelegate.java

```diff
@@ -0,0 +1,435 @@
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.zip.ZipOutputStream;
+
+import org.geoserver.config.GeoServer;
+import org.geoserver.data.util.IOUtils;
+import org.geoserver.ogr.core.Format;
+import org.geoserver.ogr.core.FormatConverter;
+import org.geoserver.ogr.core.ToolWrapper;
+import org.geoserver.ogr.core.ToolWrapperFactory;
+import org.geoserver.platform.ServiceException;
+import org.geoserver.wcs.WCSInfo;
+import org.geoserver.wcs.responses.CoverageResponseDelegate;
+import org.geotools.coverage.grid.GridCoverage2D;
+import org.geotools.coverage.grid.io.AbstractGridFormat;
+import org.geotools.gce.geotiff.GeoTiffFormat;
+import org.geotools.gce.geotiff.GeoTiffWriteParams;
+import org.geotools.gce.geotiff.GeoTiffWriter;
+import org.geotools.util.Utilities;
+import org.opengis.parameter.GeneralParameterValue;
+import org.opengis.parameter.ParameterValueGroup;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+import org.vfny.geoserver.wcs.WcsException;
+
+/**
+ * Implementation of {@link CoverageResponseDelegate} that leverages the gdal_translate utility to encode coverages in any output format supported by the
+ * GDAL library available on the system where GeoServer is running.
+ * 
+ * <p>
+ * The encoding process involves two steps:
+ * <ol>
+ * <li>the coverage passed as input to the <code>encode()</code> method is dumped to a temporary file on disk in GeoTIFF format</li>
+ * <li>the <code>gdal_translate</code> command is invoked with the provided options to convert the dumped GeoTIFF file to the desired format</li>
+ * </ol>
+ * </p>
+ * 
+ * <p>
+ * Configuration for the supported output formats must be passed to the class via its {@link #addFormat(GdalFormat)} method.
+ * </p>
+ * 
+ * @author Stefano Costa, GeoSolutions
+ */
+public class GdalCoverageResponseDelegate implements CoverageResponseDelegate, FormatConverter {
+
+    private static final GeoTiffFormat GEOTIF_FORMAT = new GeoTiffFormat();
+
+    /**
+     * Facade to GeoServer's configuration
+     */
+    GeoServer geoServer;
+
+    /**
+     * Factory to create the gdal_translate wrapper.
+     */
+    ToolWrapperFactory gdalWrapperFactory;
+
+    /**
+     * The fs path to gdal_translate. If null, we'll assume gdal_translate is in the PATH and that we can execute it just by running gdal_translate
+     */
+    String gdalTranslatePath = null;
+
+    /**
+     * The full path to gdal_translate
+     */
+    String gdalTranslateExecutable = "gdal_translate";
+
+    /**
+     * The environment variables to set before invoking gdal_translate
+     */
+    Map<String, String> environment = null;
+
+    /**
+     * Map holding the descriptors of the supported GDAL formats (keyed by format name).
+     */
+    static Map<String, Format> formats = new HashMap<String, Format>();
+    /**
+     * Map holding the descriptors of the supported GDAL formats (keyed by mime type).
```

Comment:
I'd be happy to drop the lock, but I think it's really needed, because the configuration may be updated while GeoServer is running by the [watcher](https://github.com/geoserver/geoserver/pull/1170/files#diff-512a173652411c0c49965d33fce69aaeR44) listening for changes to the `gdal_translate.xml` file.

I added a note to the docs to draw the user's attention to the fact that the configuration can be updated without a need to restart GeoServer.